aioxmpp/bookmarks/service.py

```diff
@@ -290,58 +303,107 @@ def sync(self):
         yield from self.get_bookmarks()
 
     @asyncio.coroutine
-    def add_bookmark(self, new_bookmark):
+    def add_bookmark(self, new_bookmark, max_retries=3):
         """
-        Add a bookmark.
+        Add a bookmark and check whether it was successfully added to the
+        bookmark list. Already existant bookmarks are not added twice.
 
-        Already existant bookmarks are not added twice.
+        :raises RuntimeError: if the bookmark is not in the bookmark list
+                              after `max_retries` retries.
         """
-        bookmarks = yield from self._get_bookmarks()
-        for bookmark in bookmarks:
-            if bookmark == new_bookmark:
-                break
-        else:
-            bookmarks.append(new_bookmark)
-        yield from self._set_bookmarks(bookmarks)
+        with (yield from self._lock):
```

Comment:
I think we need a ``try/finally`` with ``_diff_emit_update`` around the whole thing. Otherwise, we get inconsistent state in a listener if one of the getters/setters raises for any reason.